sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2739,6 +2752,34 @@ class Dataset[T] private[sql](
   }
 
   /**
+   * Returns true when the Dataset is cached and materialized.
+   *
+   * @group basic
+   * @since 2.3.0
+   */
+  def isMaterialized(): Boolean = {
```

Comment:
We already have a few `isCached` APIs, which mean it is cached but it might not be materialized.